source-firestore/pull.go

```diff
@@ -54,261 +39,561 @@ func (driver) Pull(stream pc.Driver_PullServer) error {
-			}
-		}
-	})
-	c.Acks = acks
-
-	// Notify Flow that we're starting.
-	log.Debug("sending PullResponse.Opened")
-	if err := c.Stream.Send(&pc.PullResponse{Opened: &pc.PullResponse_Opened{}}); err != nil {
-		return fmt.Errorf("error sending PullResponse.Opened: %w", err)
-	}
+type capture struct {
+	Requests chan *pc.PullRequest
+	Config   config
+	State    *captureState
+	Output   *captureOutput
+	Tail     bool
+}
 
-	eg.Go(func() error {
-		return c.Capture(ctx)
-	})
-	return eg.Wait()
+type captureState struct {
+	sync.RWMutex
+	Resources map[string]*resourceState
 }
 
-func (c *capture) Capture(ctx context.Context) error {
-	var scanInterval, _ = time.ParseDuration(defaultScanInterval)
-	if c.Config.ScanInterval != "" && c.Config.ScanInterval != scanIntervalNever {
-		var err error
-		scanInterval, err = time.ParseDuration(c.Config.ScanInterval)
+type resourceState struct {
+	ReadTime     time.Time
+	bindingIndex uint32
+}
 
-		if err != nil {
-			return fmt.Errorf("parsing scan interval failed: %w", err)
+// Given the prior resource states from the last DriverCheckpoint along with
+// the current capture bindings, compute a new set of resource states.
+func initResourceStates(prevStates map[string]*resourceState, resourceBindings []resource) map[string]*resourceState {
+	var states = make(map[string]*resourceState)
+	for idx, resource := range resourceBindings {
+		var state = &resourceState{bindingIndex: uint32(idx)}
+		if prevState, ok := prevStates[resource.Path]; ok {
+			state.ReadTime = prevState.ReadTime
 		}
+		states[resource.Path] = state
 	}
+	return states
+}
 
-	// Listening on snapshots gives you a copy of all documents initially
-	// so in essence it is a full scan. Hence we consider the current moment to be the last scan, and then start from there
-	var lastScan = time.Now()
+func (s *captureState) Validate() error {
+	return nil
+}
 
-	sa := option.WithCredentialsJSON([]byte(c.Config.CredentialsJSON))
-	app, err := firebase.NewApp(ctx, nil, sa)
-	if err != nil {
-		return err
+func (s *captureState) UpdateReadTimes(collectionID string, readTime time.Time) (json.RawMessage, error) {
+	s.Lock()
+	var updated = make(map[string]*resourceState)
+	for resourcePath, resourceState := range s.Resources {
+		if getLastCollectionGroupID(resourcePath) == collectionID {
+			resourceState.ReadTime = readTime
+			updated[resourcePath] = resourceState
+		}
 	}
+	s.Unlock()
 
-	client, err := app.Firestore(ctx)
+	var checkpointJSON, err = json.Marshal(&captureState{Resources: updated})
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("error serializing state checkpoint: %w", err)
 	}
-	defer client.Close()
+	return checkpointJSON, nil
+}
 
-	var eg = new(errgroup.Group)
+func (s *captureState) BindingIndex(resourcePath string) (uint32, bool) {
+	s.RLock()
+	defer s.RUnlock()
+	if state := s.Resources[resourcePath]; state != nil {
+		return state.bindingIndex, true
+	}
+	return ^uint32(0), false
```

Comment:
Took me a while to understand what this does, could use a comment to clarify it returns the maximum value of uint32

Reply:
Done